### PR TITLE
Openai generate_name_and_character_trait implementation

### PR DESCRIPTION
**Pull request by cwastche**
Refer to AGENT_CREATION_EXAMPLE for output formatting

### CODE REVIEW

**Comment by cwastche** [comment]
! Does not include a list of current npc names in the prompt to generate a unique name.
Could it be more economical to parse existing npcs and running the API call again in case of matching names?
Given that we use name + surname, it would probably happen very rarely, and thus passing a whole list to the prompt seems like a waste to me.
Thoughts ?

**Comment by edisontim** [comment]
> ! Does not include a list of current npc names in the prompt to generate a unique name. Could it be more economical to parse existing npcs and running the API call again in case of matching names? Given that we use name + surname, it would probably happen very rarely, and thus passing a whole list to the prompt seems like a waste to me. Thoughts ?

Let's first start with an implementation where we don't check names. We can decide to add this later on but for now it seems unlikely enough that it might not ever happen

**Comment by edisontim** [comment]
Do you have an estimate of this API call costs?

File: overlore/llm/constants.py [hunk 1]
```diff
@@ -82,15 +83,39 @@
 )
 
 AGENT_CREATION_TEMPLATE = """
```
Comment (edisontim) [inline overlore/llm/constants.py:85]: Can we use the same introduction sentence as for the prompt template: "You are a gamemaster for a game..." ?

File: overlore/llm/constants.py [hunk 2]
```diff
@@ -82,15 +83,39 @@
...
+     - surname
+     - {trait_type} character trait
+     - sex
+     - role
+     List of sex: {sex}
+     List of roles: {roles}
+"""
 
-    Format the speech like so: <name>:<speech>.
+AGENT_CREATION_EXAMPLE = """
```
Comment (edisontim) [inline overlore/llm/constants.py:97]: Let's name this `AGENT_CREATION_SYSTEM_PROMPT_TEMPLATE`. I think it should be 
```
Imagine you're the game master for a strategy game, you are tasked with creating Non-Playable Characters. Use these examples as reference: [enter examples]. 
Generate for an npc:
     - age
     - name
     - surname
     - character trait
     - sex (male or female)
     - role (miner, farmer, ...)
The user will tell you if the character trait needs to be positive or negative
```
I think you're mixing up what the user and system prompt needs to be. The system prompt should be the one defining what the task of the LLM is going to be. The user then uses that definition of an "assistant" to do its tasks.

Tell me how that version goes

File: overlore/llm/open_ai.py [hunk 3]
```diff
@@ -88,4 +94,20 @@ async def generate_townhall_discussion(
...
         )
 
-        return (await self.request_prompt(systemPrompt, userPrompt), systemPrompt, userPrompt)
+        return (await self.request_prompt(systemPrompt, userPrompt, "gpt-4-1106-preview"), systemPrompt, userPrompt)
+
+    async def generate_npc_profile(self) -> str:
+        "trait_type == either 'positive' or 'negative'. Otherwise, GPT calls only give out positive traits"
+        trait_type = TRAIT_TYPE[random.randrange(2)]
+
+        formatted_sex_list = ", ".join(SEX)
```
Comment (edisontim) [inline overlore/llm/open_ai.py:103]: Are the different sexes likely to change? Otherwise let's just use male or female

**Review by edisontim — CHANGES_REQUESTED** [review]

**Comment by cwastche** [comment]
> Do you have an estimate of this API call costs?

Average cost of 0.000275$
a dollar would let us generate about 3500 npcs

File: overlore/llm/open_ai.py (hunk 3 above)
Comment (cwastche) [reply overlore/llm/open_ai.py]: Probably not indeed, changed.

**Comment by cwastche** [comment]
Regarding the 'Age' generation, we might want to think about using a simple rand() function for it.
The prompt response seems to use some kind of average and keeps giving a smaller range of 'age' than asked for.
Ie, for ages between 15 and 50, it usually returns an age around the 27-35 range.

**Comment by edisontim** [comment]
> > Do you have an estimate of this API call costs?
> 
> Average cost of 0.000275$ a dollar would let us generate about 3500 npcs

Seems ok ahah

**Comment by edisontim** [comment]
> Regarding the 'Age' generation, we might want to think about using a simple rand() function for it. The prompt response seems to use some kind of average and keeps giving a smaller range of 'age' than asked for. Ie, for ages between 15 and 50, it usually returns an age around the 27-35 range.

We can see how it goes later on. I'm starting to think that an aging system where they start at age 0 and get older and older might be very fun and more immersive than a random age :)

**Review by edisontim — APPROVED** [review]
LGTM

**Comment by cwastche** [comment]
Closes #66

**Pull request merged**